Improve gb closing logic

diegoreymendez opened the pull request:
### Description:

This PR proposed an alternative fix to #996.

Unfortunately the initial proposal worked fine for a single tag, but failed when removing a tag from a larger post (sample content).

This PR proposes a more solid parsing mechanism to avoid problems with missing tags.  Opening tags without a closing match are output as regular comments.

### Details:

The solution proposed in this PR changes the delimiter parsing logic entirely, to a much more robust mechanism.

One of the things that pushed me to push this is that I realized today that the previous mechanism would have failed miserably with Gutenberg block embedding other blocks (which will be a requirement moving forward).

### Testing:

Try removing any closing or opening tag from the Gutenberg sample content.
Also try using this content in the empty editor, in HTML mode:

```
<!-- wp:list -->
<ul>
    <li>Media library/HTML for images, multimedia and approved files.</li>
    <li>Pasted links for embeds.</li>
    <li>Shortcodes for specialized assets from plugins.</li>
    <li>Featured images for the image at the top of a post or page.</li>
    <li>Excerpts for subheads.</li>
    <li>Widgets for content on the side of a page.</li>
</ul>
```

diegoreymendez commented:
Fyi, @koke

SergioEstevao commented on WordPressEditor/WordPressEditor/Classes/Plugins/WordPressPlugin/Gutenberg/GutenbergInputHTMLTreeProcessor.swift:
```diff
@@ -23,74 +26,96 @@ public class GutenbergInputHTMLTreeProcessor: HTMLTreeProcessor {
+    private func nextOpener(in nodes: ArraySlice<Node>) -> OpenerMatch? {
+        for (index, node) in nodes.enumerated() {
+            guard let commentNode = node as? CommentNode,
+                commentNode.isGutenbergBlockOpener() else {
```
I think we may need to add an extra check here for self-closing blocks.

diegoreymendez commented:
@SergioEstevao - This is ready for another look.

The changes are:

- [I fixed the self-closing block parsing logic](https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1000/files#diff-ddcfbf1973137c677397e0b64d040473R70).
- I had to merge the latest from develop, so you may see unrelated changes in this PR.  Sorry for that!
- [I removed some old tests](https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1000/files#diff-8c3b12897ab54c56754a2fff6dd0c18aL178).
- [I added a test for self-closing blocks](https://github.com/wordpress-mobile/AztecEditor-iOS/pull/1000/files#diff-8c3b12897ab54c56754a2fff6dd0c18aR129).
